test/e2e/activator_test.go

```diff
@@ -0,0 +1,193 @@
+	}
+
+	fopt := func(service *v1alpha1.Service) {
+		service.Spec.RunLatest.Configuration.RevisionTemplate.Annotations = make(map[string]string)
+		service.Spec.RunLatest.Configuration.RevisionTemplate.Annotations["autoscaling.knative.dev/maxScale"] = "10"
+	}
+
+	test.CleanupOnInterrupt(func() { TearDown(clients, helloWorldNames, logger) }, logger)
+	defer TearDown(clients, helloWorldNames, logger)
+
+	// Create a service with concurrency 1 that could sleep for N ms.
+	// Limit its maxScale to 10 containers, wait for the service to scale down and hit it with concurrent requests.
+	resources, err := test.CreateRunLatestServiceReady(logger, clients, &helloWorldNames, &configOptions, fopt)
+	if err != nil {
+		t.Fatalf("Unable to create resources: %v", err)
+	}
+	domain := resources.Route.Status.Domain
+
+	deploymentName := resources.Revision.Name + "-deployment"
+
+	logger.Info("Waiting for deployment to scale to zero.")
+	if err := pkgTest.WaitForDeploymentState(
+		clients.KubeClient,
+		deploymentName,
+		test.DeploymentScaledToZeroFunc,
+		"DeploymentScaledToZero",
+		test.ServingNamespace,
+		2*time.Minute); err != nil {
+		t.Fatalf("Failed waiting for deployment to scale to zero: %v", err)
+	}
+
+	endpoint, err := spoof.GetServiceEndpoint(clients.KubeClient.Kube)
+	if err != nil {
+		t.Fatalf("Could not get service endpoint for spoofing client")
+	}
+
+	url := fmt.Sprintf("http://%s/?timeout=%d", *endpoint, serviceSleep)
+
+	client, err := pkgTest.NewSpoofingClient(clients.KubeClient, logger, domain, test.ServingFlags.ResolvableDomain)
+
+	responseChannel := make(chan *spoof.Response, concurrency)
+	roundTrip := roundTrip(client, url)
+
+	sendRequests(roundTrip, concurrency, responseChannel, timeout, logger, t)
+
+	analyseResponses(responseChannel, concurrency, timeout, t)
+}
+
+func sendRequests(roundTrip func() (*spoof.Response, error), concurrency int, resChannel chan *spoof.Response, timeout time.Duration, logger *logging.BaseLogger, t *testing.T) {
+	var (
+		group     errgroup.Group
+		responses int32
+	)
+	doneChan := make(chan struct{})
+	timeoutChan := time.After(timeout)
+	errChan := make(chan error)
+
+	// Send out the requests asynchronously and wait for them to finish.
+	go func() {
+		logger.Info("Starting to send out the requests")
+		//	Send requests async and wait for the responses.
+		for i := 0; i < concurrency; i++ {
+			group.Go(func() error {
+				res, err := roundTrip()
+				if err != nil {
+					return fmt.Errorf("unexpected error sending a request, %v\n", err)
+				}
+				atomic.AddInt32(&responses, 1)
+				resChannel <- res
+				return nil
+			})
+		}
+		err := group.Wait()
+		if err != nil {
+			errChan <- fmt.Errorf("unexpected error making requests against activator: %v", err)
+		}
+		logger.Info("Done sending out requests")
+		doneChan <- struct{}{}
+	}()
+
+	logger.Info("Waiting for all requests to finish")
+
+done:
+	for {
+		select {
+		case <-doneChan:
+			break done
+		case <-time.Tick(5 * time.Second):
```

Comment:
This logging is the only reason why we have a label and a for loop. Does this logging really contribute that much to offset the complexity?

Reply:
I feel that having logging an output would be helpful once we need to debug them. It is also recommended in the instructions for writing the e2e tests - https://github.com/knative/serving/tree/master/test/e2e#requirements